### ISSUE TITLE
init.metadata incorrectly adds code from deferred part.

### ISSUE THREAD

**Issue by sigmundch** [dart-lang/sdk#30002]
We include in `init.metadata` information about default arguments that can be applied indirectly. When all the code regarding those default arguments are in a deferred library, we still list them in the main-chunk metadata.

Here is a simple repro:
```dart
import 'b.dart' deferred as lib;

main() => lib.loadLibrary().then((_) {
  (lib.f)();
});
```

```dart
defaultArg() => "";
f([x=defaultArg]) => x();
```

This results in a program with 2 parts, where `defaultArg` is defined in the second part, but it is listed in init.metadata on the main chunk

**Comment by sigmundch** [comment]
@floitschG @johnniwinther - I need some insights from you:
 - why do we need defaultArg in init.metadata in this case?
 - we seem to have a general issue with how we split code if it's needed in init.metadata, should we split it like we do with init.types? Any reasons not too?

**Comment by johnniwinther** [comment]
We should only include `defaultArg` in metadata if we use reflection or `Function.apply` (like https://codereview.chromium.org/2958753002) - but currently we eagerly add it.

We need to split ìnit.metadata` like `init.types` in order to support using static/top-level functions as default args in deferred libraries; otherwise we need to add functions with such default values to the main part.

**Comment by sigmundch** [comment]
Makes sense - I'm ok adding your temporary change to restrict when we add `defaultArg` for now.

**Comment by sigmundch** [comment]
Unfortunately it appears that the customer that run into this does use Function.apply, so the temporary change will not buy us much.

**Comment by sigmundch** [comment]
Fixed in ec49b67b19b241d102389e3053b760a8c12ea4ae
